### PR TITLE
Fix bugs in lambda and add map_filter function

### PR DESCRIPTION
**Pull request by haozhun**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFilterFunction.java [hunk 1]
```diff
@@ -97,8 +97,13 @@ public static Block filter(Type type, Block block, MethodHandle function)
                 throw Throwables.propagate(throwable);
             }
             if (TRUE.equals(keep)) {
-                block.writePositionTo(position, resultBuilder);
-                resultBuilder.closeEntry();
+                if (block.isNull(position)) {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFilterFunction.java]: I think `type.appendTo` would be more appropriate

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        extends SqlScalarFunction
+{
+    public static final MapFilterFunction MAP_FILTER_FUNCTION = new MapFilterFunction();
+
+    private static final MethodHandle METHOD_HANDLE = methodHandle(MapFilterFunction.class, "filter", Type.class, Type.class, Block.class, MethodHandle.class);
+
+    private MapFilterFunction()
+    {
+        super(new Signature(
+                "map_filter",
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java]: Perhaps this should just be named `filter`? @martint what do you think?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            Object value = readNativeValue(valueType, block, position + 1);
+            Boolean keep;
+            try {
+                keep = (Boolean) function.invoke(key, value);
+            }
+            catch (Throwable throwable) {
+                throw Throwables.propagate(throwable);
+            }
+            if (TRUE.equals(keep)) {
+                if (block.isNull(position)) {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java]: I think `keyType.appendTo` and `valueType.appendTo` would be more appropriate

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFilterFunction.java (hunk 1 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFilterFunction.java]: You got it. That's why I thought `writePositionTo` should suffice when I wrote it. I remember there's a way to just copy and not worried about nulls. I thought `writePositionTo` was the one.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java (hunk 2 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java]: @martint and I agreed on using different names for "overload"s that inherently different. I'll double check.

**Pull request merged**